### ISSUE TITLE
Classes with __call__ defined should be a subtype of Callable

### ISSUE THREAD

**Issue by zhangwen0411** [python/mypy#797]
```
def apply(f: Callable[[int], int], x: int) -> int:
  return f(x)

class Add5(object):
  def __call__(self, x: int) -> int:
    return x + 5

apply(Add5(), 5)
```

This code snippet doesn't type check because Mypy doesn't treat Add5() as a Callable[[int], int]

**Comment by JukkaL** [comment]
Moved to 0.5 since this comes up pretty frequently.

**Comment by margaret** [comment]
I'm going to start looking into this one.

**Comment by margaret** [comment]
WIP at https://github.com/margaret/mypy/tree/user_defined_callable
Currently debugging broken tests, specifically 
```
class A(type):
    def __init__(self) -> None: pass
def f(x: type): pass
f(A())
```
Does not work and returns
`error: Argument 1 to "f" has incompatible type Callable[[Any, Any], Any]; expected "type"`

**Comment by gvanrossum** [comment]
Have you tried putting a `pdb.set_trace()` call in `errors.Errors.report()`
and tracing back to the cause of this message?

I believe `type` is special in some ways, maybe that's biting you?

On Fri, Jun 3, 2016 at 7:33 PM, Margaret Sy <notifications@github.com>
wrote:

> WIP at https://github.com/margaret/mypy/tree/user_defined_callable
> Currently debugging broken tests, specifically
>
> class A(type):
>     def __init__(self) -> None: pass
> def f(x: type): pass
> f(A())
>
> Does not work and returns
> error: Argument 1 to "f" has incompatible type Callable[[Any, Any], Any];
> expected "type"
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/mypy/issues/797#issuecomment-223731348>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ACwrMlU02GpgXeGUJobz55IoQXJZ9oczks5qIOP8gaJpZM4FtydI>
> .
>



-- 
--Guido van Rossum (python.org/~guido)

**Comment by gvanrossum** [comment]
Moving to an earlier milestone since the partial issue is kind of important.

**Comment by JukkaL** [comment]
A related issue is the join of `Callable` and an instance type that defines `__call__` (see #2779).

**Comment by gvanrossum** [comment]
@ilevkivskyi Are you still working on this? If not it's better to un-assign.

**Comment by ilevkivskyi** [comment]
Protocols #3132 will fix this, no additional work here will be needed when they are merged.

**Comment by ilevkivskyi** [comment]
This is now fixed by #3132
